go/vt/vttablet/tabletserver/messager/message_manager.go

```diff
@@ -541,68 +553,181 @@ func (mm *messageManager) postpone(tsv TabletService, name string, ackWaitTime t
 	}
 }
 
+func (mm *messageManager) startVStream() {
+	mm.streamMu.Lock()
+	defer mm.streamMu.Unlock()
+	if mm.streamCancel != nil {
+		return
+	}
+	var ctx context.Context
+	ctx, mm.streamCancel = context.WithCancel(tabletenv.LocalContext())
+	go mm.runVStream(ctx)
+}
+
+func (mm *messageManager) stopVStream() {
+	mm.streamMu.Lock()
+	defer mm.streamMu.Unlock()
+	if mm.streamCancel != nil {
+		mm.streamCancel()
+		mm.streamCancel = nil
+	}
+}
+
+func (mm *messageManager) runVStream(ctx context.Context) {
+	for {
+		err := mm.runOneVStream(ctx)
+		select {
+		case <-ctx.Done():
+			log.Info("Context canceled, exiting vstream")
+			return
+		default:
+		}
+		MessageStats.Add([]string{mm.name.String(), "VStreamFailed"}, 1)
+		log.Infof("VStream ended: %v, retrying in 5 seconds", err)
+		time.Sleep(5 * time.Second)
+	}
+}
+
+func (mm *messageManager) runOneVStream(ctx context.Context) error {
```

Comment:
I would really appreciate some high level comments describing this event loop, and what scenarios cause rows to be skipped vs processed.

Reply:
Very helpful